libs/hooks/debounce/src/lib/debounce-hook.ts

```diff
@@ -0,0 +1,190 @@
+ */
+export class DebounceHook<T extends FlagValue = FlagValue> implements Hook {
+  private readonly cache: FixedSizeExpiringCache<true | CachedError>;
+  private readonly cacheErrors: boolean;
+
+  public constructor(
+    private readonly innerHook: Hook,
+    private readonly options: Options<T>,
+  ) {
+    this.cacheErrors = options.cacheErrors || false;
+    this.cache = new FixedSizeExpiringCache<true | CachedError>({
+      maxItems: options.maxCacheItems,
+      ttlMs: options.debounceTime,
+    });
+  }
+
+  before(hookContext: HookContext, hookHints?: HookHints) {
+    this.maybeSkipAndCache(
+      'before',
+      () => this.options?.beforeCacheKeySupplier?.(hookContext.flagKey, hookContext.context),
+      () => this.innerHook?.before?.(hookContext, hookHints),
+    );
+  }
+
+  after(hookContext: HookContext, evaluationDetails: EvaluationDetails<T>, hookHints?: HookHints) {
+    this.maybeSkipAndCache(
+      'after',
+      () => this.options?.afterCacheKeySupplier?.(hookContext.flagKey, hookContext.context, evaluationDetails),
+      () => this.innerHook?.after?.(hookContext, evaluationDetails, hookHints),
+    );
+  }
+
+  error(hookContext: HookContext, err: unknown, hookHints?: HookHints) {
+    this.maybeSkipAndCache(
+      'error',
+      () => this.options?.errorCacheKeySupplier?.(hookContext.flagKey, hookContext.context, err),
+      () => this.innerHook?.error?.(hookContext, err, hookHints),
+    );
+  }
+
+  finally(hookContext: HookContext, evaluationDetails: EvaluationDetails<T>, hookHints?: HookHints) {
+    this.maybeSkipAndCache(
+      'finally',
+      () => this.options?.finallyCacheKeySupplier?.(hookContext.flagKey, hookContext.context, evaluationDetails),
+      () => this.innerHook?.finally?.(hookContext, evaluationDetails, hookHints),
+    );
+  }
+
+  private maybeSkipAndCache(
+    stage: 'before' | 'after' | 'error' | 'finally',
+    keyGenCallback: () => string | null | undefined,
+    hookCallback: () => void,
+  ) {
+    // the cache key is a concatenation of the result of calling keyGenCallback and the stage
+    const dynamicKey = keyGenCallback();
+
+    // if the keyGenCallback returns nothing, we don't do any caching
+    if (dynamicKey) {
+      const cacheKeySuffix = stage;
+      const cacheKey = `${dynamicKey}::${cacheKeySuffix}`;
+      const got = this.cache.get(cacheKey);
+
+      if (got) {
+        // throw cached errors
+        if (got instanceof CachedError) {
+          throw got;
+        }
+        return;
+      } else {
+        try {
+          hookCallback.call(this.innerHook);
+          this.cache.set(cacheKey, true);
+        } catch (error: unknown) {
+          if (this.cacheErrors) {
+            // cache error
+            this.cache.set(cacheKey, new CachedError(error));
+          }
+          throw error;
+        }
+        return;
+      }
+    } else {
+      hookCallback.call(this.innerHook);
+    }
```

Comment:
```suggestion
    const dynamicKey = keyGenCallback();

    // if the keyGenCallback returns nothing, we don't do any caching 
    if (!dynamicKey) {
      hookCallback.call(this.innerHook);
    }
    
    const cacheKeySuffix = stage;
    const cacheKey = `${dynamicKey}::${cacheKeySuffix}`;
    const got = this.cache.get(cacheKey);

    if (got) {
      // throw cached errors
      if (got instanceof CachedError) {
        throw got;
      }
      return;
    } 
    
    try {
      hookCallback.call(this.innerHook);
      this.cache.set(cacheKey, true);
    } catch (error: unknown) {
      if (this.cacheErrors) {
        // cache error
        this.cache.set(cacheKey, new CachedError(error));
      }
      throw error;
    }    
```

This was a bit hard to read to me. Maybe we can flatten it like suggested?